### ISSUE TITLE
feat(transforms): ContinuousAffineCoupling — time-gated coupling with (time, control) conditioning

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/gauss_flows#65]
## Problem / Request

Add `ContinuousAffineCoupling` — a **time-gated** affine coupling bijection that is the identity at `t = 0` and can optionally take a **control / covariate vector** `c` alongside time `t`. Ship alongside a tiny `pack_time_control` / `unpack_time_control` helper pair so the FlowJax single-`condition` API can carry `(t, c)` cleanly.

This is the minimal first step toward **neural flows** (continuous-time invertible networks) in `gauss_flows`. It lands the condition-packing convention that subsequent neural-flow layers (FFJORD-style CNF, Continuous-iResNet, etc.) will reuse.

## User Story

> As a geophysical researcher working with irregularly-sampled or seasonally-indexed data, I want a time-conditional normalizing flow `p(x | t, c)` that evaluates likelihoods at any continuous `t` (not just observed timestamps) and is cheap enough to sweep across dense `t` grids, so that I can compute predictability / entropy / mutual-information **as functions of time** from a single trained model.

Secondary stories:

> As a researcher doing simulation-based inference, I want one flow that amortizes `p(θ | x, t)` across a family of simulation times, so that I don't have to retrain per-`t`.

> As a user studying regime-dependent distributions, I want `t` to be reinterpretable as a **physical parameter** (e.g. CO₂ level, seasonal phase) and `c` as static covariates, so that a single flow captures a continuous family of densities with identity at the reference parameter.

### Why neural flows, and why `(t, c)` conditioning

Static flows model `p(x)`; conditional flows handle `p(x | c)`; **neural flows** (Biloš et al. 2021) extend both by letting a continuous time variable `t` parameterise the transform itself, with the invariant `f(t=0, x) = x`. This unlocks problem classes the static / discretely-conditional setup cannot reach:

- **Irregularly-sampled time series** — `p(x_t | history, t)` evaluable off-grid (satellite passes, Argo floats, sensor dropouts, clinical records). The original Biloš 2021 motivation.
- **Forecasting with uncertainty** — `p(x_{t+Δ} | x_t, Δ)` with forecast horizon `Δ` as `t`. Control `c` = auxiliary covariates (forcing, boundary conditions).
- **Parameter-conditional densities** — `t` = physical parameter (temperature, phase, regime), `c` = covariates. E.g. "how does the distribution of SST anomalies deform as CO₂ increases?" The time-gate guarantees identity at the reference parameter.
- **Amortized posteriors for SBI** — `p(θ | x, t)` where `t` is a simulation time index.
- **Continuous-family information theory** — uniquely valuable because `gauss_flows` already ships `entropy` / `mutual_information` / `total_correlation`. With a time-conditional flow you get `H(X | t)`, `I(X; Y | t)`, `TC(X | t)` as **smooth functions of `t`** (predictability curves, regime-dependent dependence, MI-as-a-function-of-lead-time) from **one** trained flow.

### Why this layer first

Of the three canonical neural-flow families from stribor / Biloš 2021 (continuous-time coupling, FFJORD-style CNF, Continuous-iResNet), the time-gated affine coupling:

- has **closed-form log-det** (no ODE solver, no Hutchinson trace estimator),
- is a **minimal diff** from the existing `AffineCoupling`,
- **validates the condition-packing convention** before the heavier FFJORD / iResNet work lands,
- is cheap enough to evaluate on **dense `t` grids** — exactly what IT-measure-as-function-of-`t` needs.

So it gives us "neural flows" in `gauss_flows` at minimum cost, and unblocks the follow-up work behind one consistent `(t, c)` convention.

### Why a `pack_time_control` helper (and not just inlined concat)

FlowJax's `AbstractBijection` exposes one `condition` argument with one `cond_shape` — there is no native "structured condition" for `(t, c)`. The clean fix is to pack `[t, c]` into a single `Array` and slice inside the bijection. Centralising pack/unpack in one module:

- standardises the layout (time first, control second) across every future neural-flow layer,
- keeps FlowJax's typed `cond_shape` / `Transformed` / `Chain` / `FlowDist` / `FlowGuide` machinery working unchanged,
- makes the time-only case (`control_dim=0`) a graceful degrade,
- gives a single spot to extend (e.g. `(t, c, mask)`) without touching every bijection.

### 1. Condition-packing helpers — `src/gauss_flows/_src/conditioning.py` (new)

```python
"""Helpers for packing (time, control) into FlowJax's single-condition API."""

from __future__ import annotations

import jax.numpy as jnp
from jax import Array
from jaxtyping import ArrayLike


def pack_time_control(
    t: ArrayLike,
    control: ArrayLike | None = None,
) -> Array:
    """Pack time and optional control vector into one condition array.

    Works unbatched (t scalar, control shape (C,)) or batched on any leading
    axes (t shape (..., 1) or (...,), control shape (..., C)).

    Args:
        t: scalar or (..., 1) / (...,) time value(s).
        control: optional (..., C) control vector(s).

    Returns:
        Packed condition of shape (..., 1 + C) — or (..., 1) if control is None.
    """


def unpack_time_control(
    condition: Array,
    control_dim: int = 0,
) -> tuple[Array, Array | None]:
    """Inverse of pack_time_control. Returns (t, control_or_None).

    Args:
        condition: packed condition of shape (..., 1 + control_dim).
        control_dim: size of the control vector (0 for time-only).

    Returns:
        (t, control) where t has shape (..., 1) and control has shape
        (..., control_dim) or is None.
    """


def time_control_cond_shape(control_dim: int = 0) -> tuple[int, ...]:
    """cond_shape tuple for a bijection conditioning on (time, control)."""
```

### 2. Time-gated affine coupling — `src/gauss_flows/_src/transforms/bijections/coupling/continuous_affine_coupling.py` (new)

```python
class ContinuousAffineCoupling(AbstractBijection):
    """Time-gated affine coupling (identity at t=0, optional control vector).

    Computes
        y_a = x_a
        y_b = x_b · exp(g(t) · log_scale(x_a, c)) + g(t) · shift(x_a, c)
    where:
      - x_a / x_b are the two halves under `mask`,
      - (shift, log_scale) come from a learned NN of the passive half
        and the optional control vector `c`,
      - g(t) is a learned scalar time-gate with g(0) = 0, so the map is
        the identity at t = 0.

    cond_shape = (1 + control_dim,):  condition = concat([t, c], axis=-1).
    Use :func:`pack_time_control` to build the condition.

    Args:
        key: PRNG key for NN / time-net initialisation.
        shape: event shape (n_dims,).
        control_dim: size of the control / covariate vector c. 0 for time-only.
        n_hidden: width of the coupling MLP.
        n_layers: depth of the coupling MLP.
        time_embedding_dim: output dim of the time embedding before the gate
            (Fourier features recommended; see TimeNet).
    """

    shape: tuple[int, ...]
    cond_shape: tuple[int, ...]
    control_dim: int = eqx.field(static=True)
    mask: Array                # event-shape binary mask
    nn: eqx.Module             # (x_a [+ c]) -> (shift, log_scale)
    time_net: eqx.Module       # t -> scalar gate with g(0) = 0

    def __init__(
        self,
        key: PRNGKeyArray,
        shape: tuple[int, ...],
        *,
        control_dim: int = 0,
        n_hidden: int = 64,
        n_layers: int = 2,
        time_embedding_dim: int = 16,
    ): ...

    def transform_and_log_det(self, x, condition): ...
    def inverse_and_log_det(self, y, condition): ...
```

### 3. `TimeNet` family — `src/gauss_flows/_src/nn/time_net.py` (new module, populates the currently-empty `nn/` stub)

Minimal learnable time embeddings / gates, adapted from [stribor/net/time_net.py](https://github.com/mbilos/stribor/blob/master/stribor/net/time_net.py). All variants satisfy `g(0) = 0` **by construction** (checked in tests):

```python
class TimeFourier(eqx.Module):
    """Fourier-feature time embedding followed by a learned linear gate.

    g(t) = wᵀ sin(ω · t + φ),  with ω, φ, w learnable; initialised so
    w · sin(φ) = 0 at t = 0. Most expressive; recommended default.
    """

class TimeTanh(eqx.Module):
    """g(t) = w · tanh(a · t),  guaranteed 0 at t = 0."""

class TimeIdentity(eqx.Module):
    """g(t) = t. No parameters. Useful as a sanity baseline / fixed-time prior."""
```

### 4. Caller-side usage

```python
from gauss_flows import ContinuousAffineCoupling, pack_time_control

key = jr.key(0)
bij = ContinuousAffineCoupling(key, shape=(8,), control_dim=3)

# Single eval
cond = pack_time_control(t=0.7, control=jnp.array([1.0, 2.0, 3.0]))
y, ld = bij.transform_and_log_det(x, cond)

# Batched (vmap over time)
cond_batch = jax.vmap(lambda ti: pack_time_control(ti, c))(t_batch)

# Time-only (no control)
bij_t = ContinuousAffineCoupling(key, shape=(8,), control_dim=0)
cond = pack_time_control(0.5)   # shape (1,)

# Identity at t = 0 (held by construction)
y0, ld0 = bij.transform_and_log_det(x, pack_time_control(0.0, c))
assert jnp.allclose(y0, x, atol=1e-6) and jnp.allclose(ld0, 0.0, atol=1e-6)
```

## Design Snapshot

```python
# Inside ContinuousAffineCoupling:

def transform_and_log_det(self, x, condition):
    x = jnp.asarray(x)
    t, c = unpack_time_control(condition, self.control_dim)
    g = self.time_net(t)                       # scalar; g(0) = 0

    x_a = x * self.mask
    nn_in = x_a if c is None else jnp.concatenate([x_a, c], axis=-1)
    shift, log_scale = jnp.split(self.nn(nn_in), 2, axis=-1)

    shift, log_scale = g * shift, g * log_scale   # time-gated → identity at t=0
    y = x_a + (1 - self.mask) * (x * jnp.exp(log_scale) + shift)
    log_det = jnp.sum((1 - self.mask) * log_scale)
    return y, log_det
```

## Mathematical Notes

```text
Time-gated affine coupling:
    x_a = x ⊙ m,   x_b = x ⊙ (1 − m)
    (s, ℓ) = NN([x_a; c])                     (s = shift, ℓ = log_scale)
    g = g(t)                                  (scalar gate, g(0) = 0)
    y_a = x_a
    y_b = x_b · exp(g · ℓ) + g · s
    log|det ∂y/∂x| = ∑_{i : m_i = 0} g · ℓ_i

Identity at t = 0:
    g(0) = 0  ⇒  y = x  and  log_det = 0     (checked in tests, not just at init)

Invertibility:
    x_b = (y_b − g · s) · exp(−g · ℓ)
    exact (no fixed-point iteration needed)

Why time as a scalar gate (not additive):
    Additive  y = x + g(t) · f(x, c)  would lose the exp-scale term and
    regress to a shift-only coupling. Multiplicative on (s, ℓ) preserves
    the full affine-coupling expressivity for t ≠ 0 while still zeroing
    at t = 0.
```

## References & Existing Code

- Reference impl: [stribor/flows/coupling.py — `ContinuousAffineCoupling`](https://github.com/mbilos/stribor/blob/master/stribor/flows/coupling.py) — PyTorch
- Reference impl: [stribor/net/time_net.py](https://github.com/mbilos/stribor/blob/master/stribor/net/time_net.py) — time-net family
- Paper: Biloš et al. 2021, *Neural Flows: Efficient Alternative to Neural ODEs for Modeling Time Series*, NeurIPS. arXiv:2110.13040
- Existing bijection this mirrors: `src/gauss_flows/_src/transforms/bijections/coupling/affine.py` (`AffineCoupling`)
- Existing empty stub to populate: `src/gauss_flows/_src/nn/__init__.py`

## Implementation Steps

- [ ] Add `pack_time_control`, `unpack_time_control`, `time_control_cond_shape` in `src/gauss_flows/_src/conditioning.py`
- [ ] Add `TimeFourier`, `TimeTanh`, `TimeIdentity` in `src/gauss_flows/_src/nn/time_net.py`
- [ ] Re-export time-net classes in `src/gauss_flows/_src/nn/__init__.py`
- [ ] Add `ContinuousAffineCoupling` in `src/gauss_flows/_src/transforms/bijections/coupling/continuous_affine_coupling.py`
- [ ] Re-export in `src/gauss_flows/_src/transforms/bijections/coupling/__init__.py`
- [ ] Re-export in `src/gauss_flows/_src/transforms/bijections/__init__.py`
- [ ] Re-export in `src/gauss_flows/_src/transforms/__init__.py`
- [ ] Re-export `ContinuousAffineCoupling`, `pack_time_control`, `unpack_time_control`, `time_control_cond_shape`, `TimeFourier`, `TimeTanh`, `TimeIdentity` at the top level `src/gauss_flows/__init__.py`
- [ ] Update `__all__` in the top-level `__init__.py`

## Definition of Done

- [ ] Code lands at the intended paths
- [ ] Public API exported via `src/gauss_flows/__init__.py`
- [ ] Tests pass: `make test`
- [ ] Lint + typecheck pass: `make lint && make typecheck`
- [ ] Google-style docstrings on every public symbol

## Testing

- [ ] **Identity at t=0**: `transform_and_log_det(x, pack_time_control(0.0, c))` returns `(x, 0.0)` to atol 1e-6 for random `x`, `c`.
- [ ] **Round-trip**: `inverse_and_log_det(transform_and_log_det(x, cond)[0], cond)[0] ≈ x` to atol 1e-5, for both `control_dim=0` and `control_dim>0`.
- [ ] **Log-det consistency**: forward log-det + inverse log-det ≈ 0 to atol 1e-5.
- [ ] **Autodiff agreement**: computed log-det matches `jnp.linalg.slogdet(jax.jacrev(...))` on small (dim ≤ 6) random examples.
- [ ] **JIT / grad safety**: `eqx.filter_jit(bij.transform_and_log_det)` matches eager; `eqx.filter_grad(loss)(bij)` returns finite gradients with non-zero entries for `time_net` and `nn` params.
- [ ] **pack/unpack round-trip**: for `control_dim ∈ {0, 1, 3, 8}`, `unpack_time_control(pack_time_control(t, c), control_dim)` recovers the inputs (modulo the (…, 1) reshape of t).
- [ ] **Time-only path**: `control_dim=0` works end-to-end and `c is None` inside `transform_and_log_det`.
- [ ] **Batched vmap**: `jax.vmap(bij.transform_and_log_det, in_axes=(0, 0))` matches per-example evaluation.
- [ ] **TimeNet gate test**: each TimeNet variant returns exactly 0 at t=0 after random init and after a few gradient steps.
- [ ] **Simple training sanity**: one tiny training loop (≤ 200 steps, 2D toy data) drives NLL down without NaNs; identity-at-t=0 still holds post-training.

## Documentation

- [ ] Docstrings (covered by Definition of Done)
- [ ] API reference entry for `ContinuousAffineCoupling`, `pack_time_control`, `unpack_time_control`, `TimeFourier`, `TimeTanh`
- [ ] Notebook / recipe demonstrating IT-measures-as-function-of-t on a toy problem — **deferred to a follow-up docs issue**

## Out of scope (explicit)

- FFJORD-style CNF (`ContinuousTransform`) — separate issue; needs `diffrax` + Hutchinson trace via `matfree`.
- Continuous-iResNet — separate issue; needs spectral norm + fixed-point inverse.
- `MatrixExponential` linear flow — separate issue; trivial port but orthogonal to this one.
- `NeuralFlow` top-level chain wrapper — defer; the standard FlowJax `Chain` / `Transformed` already composes time-conditional bijections as long as every layer shares the same `cond_shape`.
- Structured / PyTree conditions beyond the packed `Array` — rejected in this issue; revisit only if a concrete use case needs per-layer heterogeneous condition shapes.

## Estimate

~250 LOC implementation (helpers + time-net + coupling) + ~200 LOC tests. Small-to-medium PR.

## Relationships

- Parent (theme epic): TBD — neural flows / continuous-time research theme
- Blocks: follow-up issues for FFJORD CNF and Continuous-iResNet (both will reuse `pack_time_control` and the `cond_shape` convention)
- Related: scout notes on stribor (Biloš 2021 neural flows); this issue captures the first recommended port from that scout
